persistence/src/main/scala/hmda/persistence/processing/HmdaFileValidator.scala

```diff
@@ -0,0 +1,199 @@
+  }
+
+  case class HmdaFileValidationState(
+      lars: Seq[LoanApplicationRegister] = Nil,
+      syntactical: Seq[ValidationError] = Nil,
+      validity: Seq[ValidationError] = Nil,
+      quality: Seq[ValidationError] = Nil
+  ) {
+    def updated(event: Event): HmdaFileValidationState = event match {
+      case larValidated @ LarValidated(lar) =>
+        HmdaFileValidationState(lars :+ lar, syntactical, validity, quality)
+      case SyntacticalError(e) =>
+        HmdaFileValidationState(lars, syntactical :+ e, validity, quality)
+      case ValidityError(e) =>
+        HmdaFileValidationState(lars, syntactical, validity :+ e, quality)
+      case QualityError(e) =>
+        HmdaFileValidationState(lars, syntactical, validity, quality :+ e)
+
+    }
+  }
+}
+
+class HmdaFileValidator(submissionId: String) extends PersistentActor with ActorLogging with LarEngine with LocalEventPublisher {
+
+  import HmdaFileValidator._
+
+  implicit val system = context.system
+  implicit val ec = system.dispatcher
+  implicit val materializer = ActorMaterializer()
+
+  val parserPersistenceId = s"${HmdaFileParser.name}-$submissionId"
+
+  var state = HmdaFileValidationState()
+
+  def updateState(event: Event): Unit = {
+    state = state.updated(event)
+  }
+
+  override def preStart(): Unit = {
+    log.debug(s"Validation started for $submissionId")
+  }
+
+  override def postStop(): Unit = {
+    log.debug(s"Validation ended for $submissionId")
+  }
+
+  override def persistenceId: String = s"$name-$submissionId"
+
+  override def receiveCommand: Receive = {
+
+    case BeginValidation =>
+      val validationStarted = ValidationStarted(submissionId)
+      publishEvent(validationStarted)
+      self ! ValidateLarSyntactical
+
+    case ValidateLarSyntactical =>
+      events(parserPersistenceId)
+        .filter(x => x.isInstanceOf[LarParsed])
+        .map(e => e.asInstanceOf[LarParsed].lar)
+        .map(lar => checkSyntactical(lar, ValidationContext(None)).toEither)
+        .map {
+          case Right(lar) => lar
+          case Left(errors) => ValidationErrors(errors.list.toList)
+        }
+        .runForeach { x =>
+          self ! x
+        }
+        .andThen {
+          case _ => self ! ValidateLarValidity
+        }
+
+    case ValidateLarValidity =>
+      events(parserPersistenceId)
+        .map { case LarParsed(lar) => lar }
+        .map(lar => checkValidity(lar, ValidationContext(None)).toEither)
+        .map {
+          case Right(lar) => lar
+          case Left(errors) => ValidationErrors(errors.list.toList)
+        }
+        .runForeach { x =>
+          self ! x
+        }
+        .andThen {
+          case _ =>
+            self ! CompleteSyntacticalAndValidity
+            if (state.syntactical.isEmpty && state.validity.isEmpty) {
+              self ! ValidateLarQuality
+            }
+        }
+
+    case ValidateLarQuality =>
+      events(parserPersistenceId)
+        .map { case LarParsed(lar) => lar }
```

Comment:
I just realized there is a problem with this implementation. Working on an alternative solution